streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java

```diff
@@ -202,21 +210,33 @@ private void restorePartition(final Map<TopicPartition, Long> endOffsets,
                     restorer.startingOffset(),
                     restorer.restoredOffset());
 
+            restorer.restoreDone();
+
             partitionIterator.remove();
         }
     }
 
-    private long processNext(final List<ConsumerRecord<byte[], byte[]>> records, final StateRestorer restorer, final Long endOffset) {
+    private long processNext(final List<ConsumerRecord<byte[], byte[]>> records,
+                             final StateRestorer restorer, final Long endOffset) {
+        final List<KeyValue<byte[], byte[]>> restoreRecords = new ArrayList<>();
+
         for (final ConsumerRecord<byte[], byte[]> record : records) {
             final long offset = record.offset();
             if (restorer.hasCompleted(offset, endOffset)) {
+                if (!restoreRecords.isEmpty()) {
```

Comment:
This logic seems a bit complex to me, and also if we return at line 229 `restoreBatchCompleted` is not called as well. Is this correct? How about:

```
restoreRecords = new list..
nextPosition = -1;
for (...) {
    if (restorer.hasCompleted) {
        nextPosition = record.offset();
        break;
    } else {
        restoreRecords.add(...);
    }
}

if (nextPosition == -1)
    nextPosition = consumer.position(restorer.partition());

if (!restoreRecords.isEmpty()){
    restorer.restore(restoreRecords);
    restorer.restoreBatchCompleted(currentPosition, records.size());
}

return nextPosition;
```